src/sqlancer/tidb/oracle/TiDBCERTOracle.java

```diff
@@ -0,0 +1,285 @@
+            select.clearHavingClause();
+        } else {
+            select.setGroupByExpressions(select.getFetchColumns());
+        }
+        return increase;
+    }
+
+    private boolean mutateHaving() {
+        if (select.getGroupByExpressions().size() == 0) {
+            select.setGroupByExpressions(select.getFetchColumns());
+            select.setHavingClause(gen.generateExpression());
+            return false;
+        } else {
+            if (select.getHavingClause() == null) {
+                select.setHavingClause(gen.generateExpression());
+                return false;
+            } else {
+                select.setHavingClause(null);
+                return true;
+            }
+        }
+    }
+
+    private boolean mutateAnd() {
+        if (select.getWhereClause() == null) {
+            select.setWhereClause(gen.generateExpression());
+        } else {
+            TiDBExpression newWhere = new TiDBBinaryLogicalOperation(select.getWhereClause(), gen.generateExpression(),
+                    TiDBBinaryLogicalOperator.AND);
+            select.setWhereClause(newWhere);
+        }
+        return false;
+    }
+
+    private boolean mutateOr() {
+        if (select.getWhereClause() == null) {
+            select.setWhereClause(gen.generateExpression());
+            return false;
+        } else {
+            TiDBExpression newWhere = new TiDBBinaryLogicalOperation(select.getWhereClause(), gen.generateExpression(),
+                    TiDBBinaryLogicalOperator.OR);
+            select.setWhereClause(newWhere);
+            return true;
+        }
+    }
+
+    private boolean mutateLimit() {
+        boolean increase = select.getLimitClause() != null;
+        if (increase) {
+            select.setLimitClause(null);
+        } else {
+            select.setLimitClause(gen.generateConstant(TiDBDataType.INT));
+        }
+        return increase;
+    }
+
+    private double getRow(SQLGlobalState<?, ?> globalState, String selectStr, List<String> queryPlanSequences)
+            throws AssertionError, SQLException {
+        double row = -1;
+        String explainQuery = "EXPLAIN " + selectStr;
+
+        // Log the query
+        if (globalState.getOptions().logEachSelect()) {
+            globalState.getLogger().writeCurrent(explainQuery);
+            try {
+                globalState.getLogger().getCurrentFileWriter().flush();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+
+        // Get the row count
+        SQLQueryAdapter q = new SQLQueryAdapter(explainQuery, errors);
+        try (SQLancerResultSet rs = q.executeAndGet(globalState)) {
+            if (rs != null) {
+                while (rs.next()) {
+                    double estRows = Double.parseDouble(rs.getString(2));
+                    if (row == -1) {
+                        row = estRows;
+                    }
+                    String operation = rs.getString(1).split("_")[0];
```

Comment:
perhaps a comment would make clearer what we are extracting here?

Reply:
Solved